cis194/week2/robert-chang/test/LibSpec.hs

```diff
@@ -16,8 +23,23 @@ spec = do
    it "return correct unKnown" $ do
       parseMessage "haha Steve Mao" `shouldBe` Unknown "haha Steve Mao"
 
-  -- I don't know how to test this guy
-  -- describe "parse" $ do
-  --  it "returns correct result" $ do   
-     -- testParse parse 1 "test/sample.log" `shouldBe` LogMessage Info 6 "Completed armadillo processing"
-    
+insertSpec :: Spec
+insertSpec = do
+  describe "insert" $ do
+  let tree = Node (Node Leaf (LogMessage Info 29 "la la la") Leaf) (LogMessage Warning 29 "la la la") Leaf 
+  it "should insert LogMessage into the tree" $
+    insert (LogMessage (Error 2) 562 "help help") Leaf `shouldBe` Node Leaf (LogMessage (Error 2) 562 "help help") Leaf
+
+  it "should insert LogMessage into the left tree when timestamp is smaller" $
+    insert (LogMessage (Error 2) 10 "help help") tree `shouldBe` Node (Node (Node Leaf (LogMessage (Error 2) 10 "help help") Leaf) (LogMessage Info 29 "la la la") Leaf) (LogMessage Warning 29 "la la la") Leaf 
+
+  it "should insert LogMessage into the right tree when timestamp is bigger" $
+    insert (LogMessage (Error 2) 590 "help help") tree `shouldBe` Node (Node Leaf (LogMessage Info 29 "la la la") Leaf) (LogMessage Warning 29 "la la la") (Node Leaf (LogMessage (Error 2) 590 "help help") Leaf)
```

Comment:
My own two cents here (as a discussion) - tests serves mainly 2 purposes:

* An automated fundamental regression test against any code changes.
* As a documentation.

To me personally, this particular test case is beyond something that I can mentally process by just reading it. Unfortunately I do not have a good answer for this right now. I may think of writing an _unsorted_ version of `MessageTree -> [LogMessage]` (which should be trivial) as a test helper, that would increase the readability in this case. It's a slight compromise, I do understand that.